nova/scheduler/filters/shard_filter.py

```diff
@@ -148,18 +204,58 @@ def host_passes(self, host_state, spec_obj):
         if self._ALL_SHARDS in shards:
             LOG.debug('project enabled for all shards %(project_shards)s.',
                       {'project_shards': shards})
-            return True
         elif host_shard_names & set(shards):
             LOG.debug('%(host_state)s shard %(host_shard)s found in project '
                       'shards %(project_shards)s.',
                       {'host_state': host_state,
                        'host_shard': host_shard_names,
                        'project_shards': shards})
-            return True
         else:
             LOG.debug('%(host_state)s shard %(host_shard)s not found in '
                       'project shards %(project_shards)s.',
                       {'host_state': host_state,
                        'host_shard': host_shard_names,
                        'project_shards': shards})
             return False
+
+        if not utils.request_is_resize(spec_obj):
+            # K8S orchestrators are only creating or deleting nodes,
+            # therefore we shouldn't infer with resize/migrate requests.
+            return self._host_passes_k8s(host_state, host_shard_names,
+                                         spec_obj, k8s_instances)
+
+        return True
+
+    def _host_passes_k8s(self, host_state, host_shard_names, spec_obj,
+                         k8s_instances):
+        """Instances of a K8S cluster must end up on the same shard.
+        The K8S cluster is identified by the metadata or tags set
+        by the orchestrator (Gardener or Kubernikus).
+        """
+        if not k8s_instances:
+            # There are no instances in the cluster, yet.
+            # We allow any shard for the first instance.
+            return True
+
+        def _is_hana(flavor):
+            return flavor.name.startswith(HANA_PREFIX)
+
+        def _is_same_category(instance, flavor):
+            """Check whether instance is from the flavor's family."""
+            if _is_hana(flavor):
+                return _is_hana(instance.flavor)
+            return True
+
+        def _instance_matches(instance):
+            if spec_obj.availability_zone:
+                if (instance.availability_zone !=
+                        spec_obj.availability_zone):
+                    return False
+            return _is_same_category(instance, spec_obj.flavor)
+
+        k8s_hosts = set([i.host for i in k8s_instances
+                         if _instance_matches(i)])
+
+        return any(agg.name in host_shard_names and
+                   set(agg.hosts) & k8s_hosts
+                   for agg in host_state.all_aggregates)
```

Comment:
`all_aggregates` is the same dict for all the hosts - would it make sense to only compute the `set()` once for all?

Do we even need `all_aggregates` here, if we match only for aggregates our host is in via `agg.name in host_shard_names`?

Reply:
ah, so the `self.aggregates` doesn't contain the `hosts` attribute? Because it looked to me like the whole `Aggregate` object would passed along there, too.